src/core/geometry/qgscircularstring.cpp

```diff
@@ -1592,6 +1594,118 @@ double QgsCircularString::segmentLength( QgsVertexId startVertex ) const
   return QgsGeometryUtilsBase::circleLength( x1, y1, x2, y2, x3, y3 );
 }
 
+double QgsCircularString::distanceBetweenVertices( QgsVertexId fromVertex, QgsVertexId toVertex ) const
+{
+  // Convert QgsVertexId to simple vertex numbers for curves (single ring, single part)
+  if ( fromVertex.part != 0 || fromVertex.ring != 0 || toVertex.part != 0 || toVertex.ring != 0 )
+    return -1.0;
+
+  int fromVertexNumber = fromVertex.vertex;
+  int toVertexNumber = toVertex.vertex;
+
+  if ( fromVertexNumber < 0 || fromVertexNumber >= numPoints() || toVertexNumber < 0 || toVertexNumber >= numPoints() )
+    return -1.0;
+
+  if ( fromVertexNumber == toVertexNumber )
+    return 0.0;
+
+  // Ensure fromVertex < toVertex for simplicity
+  if ( fromVertexNumber > toVertexNumber )
```

Comment:
I'm not sure how often this case will eventuate, but it would be more efficient to move this check up to the top of the function so that we don't do the sanity checks twice when this does happen.

Reply:
done